src/Symfony/Component/Lock/Store/MongoDbStore.php

```diff
@@ -0,0 +1,426 @@
+use Symfony\Component\Lock\Exception\LockExpiredException;
+use Symfony\Component\Lock\Exception\LockStorageException;
+use Symfony\Component\Lock\Exception\NotSupportedException;
+use Symfony\Component\Lock\Key;
+use Symfony\Component\Lock\StoreInterface;
+
+/**
+ * MongoDbStore is a StoreInterface implementation using MongoDB as a storage
+ * engine.
+ *
+ * CAUTION: For MongoDB servers ^2.2 TTL Indexes are used so this store relies
+ * on all client and server nodes to have synchronized clocks for lock expiry
+ * to occur at the correct time. For MongoDB servers <2.2 this store relies
+ * on all client nodes to have synchronized clocks for lock expiry to occur at
+ * the correct time. To ensure locks don't expire prematurely; the TTLs should
+ * be set with enough extra time to account for any clock drift between nodes.
+ *
+ * CAUTION: The locked resource name is indexed in the _id field of the
+ * lock collection. An indexed field's value in MongoDB can be a maximum
+ * of 1024 bytes in length inclusive of structural overhead.
+ *
+ * @see https://docs.mongodb.com/manual/reference/limits/#Index-Key-Limit
+ *
+ * @requires extension mongodb
+ *
+ * @author Joe Bennett <joe@assimtech.com>
+ */
+class MongoDbStore implements StoreInterface
+{
+    private $mongo;
+    private $options;
+    private $initialTtl;
+
+    private $collection;
+    private $databaseVersion;
+
+    /**
+     * Returns whether or not the store is supported.
+     *
+     * @return bool
+     *
+     * @internal
+     */
+    public static function isSupported(): bool
+    {
+        return \extension_loaded('mongodb');
+    }
+
+    /**
+     * @param Client $mongo
+     * @param array  $options    See below
+     * @param float  $initialTtl The expiration delay of locks in seconds
+     *
+     * @throws InvalidArgumentException if required options are not provided
+     *
+     * Options:
+     *      database:     The name of the database [required]
+     *      collection:   The name of the collection [default: lock]
+     *      gcProbablity: Should a TTL Index be created expressed as a probability from 0.0 to 1.0 [default: 0.001]
+     *
+     * If gcProbablity is set to a value greater than 0.0 there is a chance
+     * this store will attempt to create a TTL index on self::save().
+     * If however the server version is less than MongoDB 2.2 indicating TTL
+     * Indexes are unsupported, a db.lock.remove() query will be used to
+     * cleanup old locks instead. If you prefer to create your TTL Index
+     * manually you can set gcProbablity to 0.0 and optionally leverage
+     * self::createTtlIndex(int $expireAfterSeconds = 0).
+     *
+     * writeConcern, readConcern and readPreference are not specified by
+     * MongoDbStore meaning the collection's settings will take effect.
+     *
+     * @see https://docs.mongodb.com/manual/applications/replication/
+     */
+    public function __construct(Client $mongo, array $options, float $initialTtl = 300.0)
```

Comment:
you should not type-hint the mongodb client, what happens if he does not exist ?

Reply:
Then the developer wouldn't be able to construct a `MongoDB\Client`. If they pass something that isn't a `MongoDB\Client` they would get a TypeError:

```
Uncaught TypeError: Argument 1 passed to Symfony\Component\Lock\Store\MongoDbStore::__construct() must be an instance of MongoDB\Client, x given
```

This is what I would want to happen

Reply:
What happens if you don't have MongoDB ?

Reply:
if you don't have `mongodb/mongodb` you can't construct a `MongoDB\Client`

if you don't have `ext-mongodb` you can't `composer require mongodb/mongodb`

if you somehow have installed `mongodb/mongodb` and don't have `ext-mongodb` you will file to construct a `MongoDB\Client` (this is `mongodb/mongodb`'s responsibility):

```
$ docker run --rm -it -v $PWD:$PWD -w $PWD php php ./test.php

Fatal error: Uncaught Error: Class 'MongoDB\Driver\Manager' not found in /home/kralos/src/symfony/vendor/mongodb/mongodb/src/Client.php:87
Stack trace:
#0 /home/kralos/src/symfony/test.php(14): MongoDB\Client->__construct('mongodb://127.0...')
#1 {main}
  thrown in /home/kralos/src/symfony/vendor/mongodb/mongodb/src/Client.php on line 87
```

Reply:
`MongoDB\Driver\Manager` is provided by `ext-mongodb`

`MongoDB\Client` is provided by `mongodb/mongodb`

`Symfony\Component\Lock\Store\MongoDbStore` depends on `mongodb/mongodb`

Reply:
I think you not type-hint MongoDB\Client for that reason. maybe i'm wrong cc @jderusse

Reply:
I removed the `if (!\extension_loaded('mongodb')) {` check in the constructor since it's impossible to hit it. You can't pass a `MongoDB\Client` without `mongodb/mongodb` which you can't have without `ext-mongodb`